langchain4j/src/main/java/dev/langchain4j/service/tool/ToolProviderResult.java

```diff
@@ -0,0 +1,22 @@
+package dev.langchain4j.service.tool;
+
+import dev.langchain4j.agent.tool.ToolSpecification;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A wrapper mostly used by the {@link ToolProvider}
+ */
+public class ToolProviderResult {
+    private final Map<ToolSpecification, ToolExecutor> tools = new HashMap<>();
+
+    public Map<ToolSpecification, ToolExecutor> getTools() {
+        return Collections.unmodifiableMap(tools);
+    }
+
+    public void add(ToolSpecification specification, ToolExecutor toolExecutor) {
```

Comment:
All our domain objects are immutable, please change it so that all tools are provided via constructor

Reply:
Sure. I would change the constructor to this:
```Java
public ToolProviderResult(Map<ToolSpecification, ToolExecutor> tools) {
        this.tools = tools;
}
```

But then we get more boilerplate code.
For example:
```JAVA
Map<ToolSpecification, ToolExecutor> toolMap = new HashMap<>();
toolMap.put(specification, executor);
ToolProviderResult bookingResult = new ToolProviderResult(toolMap);
 ```
 
Is this okay, or do we want to improve this?
I thought of something with varargs, but this didn't work as expected.

Reply:
What about:
```java
ToolProviderResult.builder()
    .add(spec1, executor1)
    .add(spec2, executor2)
    ...
    .build();
```

But there should be a `.add(Map<Spec,Executor>)` as well.

Reply:
Update: I have just discussed this with @d135-1r43, and he thinks it's okay to write a little bit more boilerplate code.

Reply:
The builder pattern is a way cleaner solution. I have added your idea and @d135-1r43 reviewed my code.
You find my commit [here](https://github.com/MiggiV2/langchain4j/commit/c80710140c4242be7bb07d616c9aa862a76996a0).
WDYT?